### PR TITLE
Complete rewrite using reanimated v2

### PR DESCRIPTION
**Pull request by thehobbit85**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1187780754031539

### CODE REVIEW

**Comment by memoalv** [comment]
any chance this is usable?

File: src/alphabets/charSets.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+
+export const Digits = '0123456789'
+export const DecimalPoints = ',.'
```
Comment (samholmes) [inline src/alphabets/charSets.ts]: `NumberDelimiters`?

File: src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+const Tick = (props: TickProps) => {
+  const {
+    children, duration = 500, rotateItems = [], tickerType = 'spring',
+  } = props
+
+  const height = useSharedValue(props.measureMap.height)
+  const width = useSharedValue(props.measureMap.width)
+
+  const rotateItemsStyle = useAnimatedStyle(() => {
+    'worklet'
```
Comment (samholmes) [inline src/index.tsx:40]: `'worklet'` directive can be omitted for `useAnimatedStyle` hook.

File: src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    const toHeight = rotateItems.indexOf(children) * height.value * -1
```
Comment (samholmes) [inline src/index.tsx]: Is this a height? Seems like a Y coordinate. You sure this shouldn't be named `toY` or `destinationY`.

File: src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  })
+
+  return (
+    <Animated.View style={[viewPortStyle]} >
+      <Animated.View style={[rotateItemsStyle]} >
+        {rotateItems.map((text, i) => (
+          <Animated.Text key={i} {...props} style={[props.textStyle, { height: height.value }]}>
+            {text}
+          </Animated.Text>
```
Comment (samholmes) [inline src/index.tsx]: Hmm, is react complaining about the key set to an index here? If not, then no worries.

File: src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  ), [ measureMap ])
+
+  const childProps = { ...props, rotateItems }
+  return (
+    <View style={styles.row}>
+      {hiddenChildren}
+      {!measured
+        ? null
+        : textInput.map((char, i) =>
+          <Tick {...childProps} measureMap={measureMap[char] ?? {}} key={i}>
```
Comment (samholmes) [inline src/index.tsx]: Why the `?? {}`? Seems like `measureMap[char]` should always be defined at this point.

File: src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+export const Ticker = (props: TickerProps) => {
+  const [measured, setMeasured] = useState<boolean>(false)
+  const measureMap = useRef<{ [key: string]: MeasureMap }>({}).current
+  const textInput = `${props.value ?? props.children[0]}`.split('')
+
+  const rotateItems: string[] = props.rotateItems != null
+    ? [ ...new Set(props.rotateItems.join('').split('')) ]
+    : getAlphabet(RotateAlphabets, props.charSet ?? 'Keypad.Numpad')
```
Comment (samholmes) [inline src/index.tsx]: If `props.rotateItem` mutates, does this mutate the `measureMap`? Where do we do `setMeasured(false)` when this happens?

Comment (samholmes) [reply src/index.tsx]: Perhaps we need to add an effect which uses rotateItems as a dep.

**Review by samholmes — CHANGES_REQUESTED** [review]

**Comment by thehobbit85** [comment]
> any chance this is usable?

Yes it is. But the documentation is completely useless now since the new implementation has a very different API. You’re more then welcome to try and figure out the API from the code or you can wait a bit until I update the docs.

File: src/index.tsx (hunk 2 above)
Comment (thehobbit85) [reply src/index.tsx:40]: In theory, yes. But for some reason when I tried the App exploded. Not sure why but that's why I kept them.

File: src/index.tsx (hunk 3 above)
Comment (thehobbit85) [reply src/index.tsx]: Changed into "toTranslateY"

File: src/index.tsx (hunk 5 above)
Comment (thehobbit85) [reply src/index.tsx]: You are correct

File: src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  )
+}
+
+export const Ticker = (props: TickerProps) => {
+  const [measured, setMeasured] = useState<boolean>(false)
+  const measureMap = useRef<{ [key: string]: MeasureMap }>({}).current
+  const textInput = `${props.value ?? props.children[0]}`.split('')
+
+  const rotateItems: string[] = props.rotateItems != null
+    ? [ ...new Set(props.rotateItems.join('').split('')) ]
```
Comment (samholmes) [inline src/index.tsx]: This `.join('').split('')` is a cause of concern.

File: src/index.tsx [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  const hiddenChildren = useMemo(() => allCharacters.map((char, i) =>
+    <Text
+      {...props}
+      style={[props.textStyle, styles.hide]}
+      key={i}
+      onLayout={onLayout(char)}
+    >{char}</Text>
+  ), [ measureMap ])
```
Comment (samholmes) [inline src/index.tsx]: Ah. Move the `if (measureMap == null) return` check into the memoization function.

```ts
  const onLayout = (char: string) => (event: any) => {
    measureMap[char] = { ...event?.nativeEvent?.layout }
    if (Object.keys(measureMap).length === allCharacters.length) {
      setMeasured(true)
    }
  }

  const hiddenChildren = useMemo(() => allCharacters.map((char, i) => {
    if (measureMap == null) return null
    return <Text
      {...props}
      style={[props.textStyle, styles.hide]}
      key={i}
      onLayout={onLayout(char)}
    >{char}</Text>
  }), [ measureMap ])
```

This way you only create the children once the `measureMap` ref is ready.

**Review by samholmes — APPROVED** [review]
Approved on final minor change.

File: src/index.tsx (hunk 8 above)
Comment (thehobbit85) [reply src/index.tsx]: I like that idea, but you got the code a bit wrong, this is what I changed it into:

```js
 const onLayout = (char: string) => (event: any) => {
    measureMap[char] = { ...event?.nativeEvent?.layout }
    if (Object.keys(measureMap).length === allCharacters.length) {
      setMeasured(true)
    }
  }

  const hiddenChildren = useMemo(() => {
    if (measureMap == null) return
    return allCharacters.map((char, i) =>
      <Text
        {...props}
        style={[props.textStyle, styles.hide]}
        key={i}
        onLayout={onLayout(char)}
      >{char}</Text>)
  }, [ measureMap ])
```

**Pull request merged**